File: .net/vb.net/Server/ServerMain.vb
Public Class fmServerMain
    Private FManager As wclBluetoothManager
    Private FServer As wclRfCommServer

#Region "Helper methods."
    Private Overloads Sub Trace(str As String)
        lbLog.Items.Add(str)
    End Sub

    Private Overloads Sub Trace(str As String, [Error] As Int32)
        Trace(str + ": &H" + [Error].ToString("X8"))
    End Sub
#End Region

#Region "Bluetooth Manager events."
    Private Sub FManager_BeforeClose(sender As Object, e As EventArgs)
        Trace("Bluetooth Manager closed")
    End Sub

    Private Sub FManager_AfterOpen(sender As Object, e As EventArgs)
        Trace("Bluetooth Manager opened.")
    End Sub
#End Region

#Region "Server events."
    Private Sub FServer_OnListen(sender As Object, e As EventArgs)
        Trace("Server listening")
    End Sub

    Private Sub FServer_OnDisconnect(Sender As Object, Client As wclRfCommServerClientConnection, Reason As Int32)
        Trace("Client disconnected", Reason)
    End Sub

    Private Sub FServer_OnDestroyProcessor(Sender As Object, Connection As wclServerClientDataConnection)
        If Connection.Processor IsNot Nothing Then
            Trace("Destroy data processor")
            Connection.Processor.Dispose()
        End If
    End Sub

    Private Sub FServer_OnCreateProcessor(Sender As Object, Connection As wclServerClientDataConnection)
        Trace("Creating data processor")
        Dim Proc As ServerDataProcessor = New ServerDataProcessor(Connection)

        AddHandler Proc.OnByteReceived, AddressOf ByteReceived
        AddHandler Proc.OnUInt16Received, AddressOf UInt16Received
        AddHandler Proc.OnUInt32Received, AddressOf UInt32Received
        AddHandler Proc.OnUInt64Received, AddressOf UInt64Received
        AddHandler Proc.OnSByteReceived, AddressOf SByteReceived
        AddHandler Proc.OnInt16Received, AddressOf Int16Received
        AddHandler Proc.OnInt32Received, AddressOf Int32Received
        AddHandler Proc.OnInt64Received, AddressOf Int64Received
        AddHandler Proc.OnArrayReceived, AddressOf ArrayReceived
        AddHandler Proc.OnStringReceived, AddressOf StringReceived

        AddHandler Proc.OnGetByte, AddressOf GetByte
        AddHandler Proc.OnGetUInt16, AddressOf GetUInt16
        AddHandler Proc.OnGetUInt32, AddressOf GetUInt32
        AddHandler Proc.OnGetUInt64, AddressOf GetUInt64
        AddHandler Proc.OnGetSByte, AddressOf GetSByte
        AddHandler Proc.OnGetInt16, AddressOf GetInt16
        AddHandler Proc.OnGetInt32, AddressOf GetInt32
        AddHandler Proc.OnGetInt64, AddressOf GetInt64
        AddHandler Proc.OnGetArray, AddressOf GetArray
        AddHandler Proc.OnGetString, AddressOf GetString
    End Sub

    Private Sub FServer_OnConnect(Sender As Object, Client As wclRfCommServerClientConnection, [Error] As Int32)
        Trace("Client connect", [Error])
    End Sub

    Private Sub FServer_OnClosed(Sender As Object, Reason As Int32)
        Trace("Server closed")
    End Sub
#End Region

#Region "Data processor events."
#Region "Data received."
    Private Sub ByteReceived(Sender As Object, Data As Byte)
        Trace("Byte received: " + Data.ToString("X2") + " (" + Data.ToString() + ")")
    End Sub

    Private Sub UInt16Received(Sender As Object, Data As UInt16)
        Trace("UInt16 received: " + Data.ToString("X4") + " (" + Data.ToString() + ")")
    End Sub

    Private Sub UInt32Received(Sender As Object, Data As UInt32)
        Trace("UInt32 received: " + Data.ToString("X8") + " (" + Data.ToString() + ")")
    End Sub

    Private Sub UInt64Received(Sender As Object, Data As UInt64)
        Trace("UInt64 received: " + Data.ToString("X16") + " (" + Data.ToString() + ")")
    End Sub

    Private Sub SByteReceived(Sender As Object, Data As SByte)
        Trace("SByte received: " + Data.ToString("X2") + " (" + Data.ToString() + ")")
    End Sub

    Private Sub Int16Received(Sender As Object, Data As Int16)
        Trace("Int16 received: " + Data.ToString("X4") + " (" + Data.ToString() + ")")
    End Sub

    Private Sub Int32Received(Sender As Object, Data As Int32)
        Trace("Int32 received: " + Data.ToString("X8") + " (" + Data.ToString() + ")")
    End Sub

    Private Sub Int64Received(Sender As Object, Data As Int64)
        Trace("Int64 received: " + Data.ToString("X16") + " (" + Data.ToString() + ")")
    End Sub

    Private Sub ArrayReceived(Sender As Object, Data As Byte())
        Trace("Array received: " + Data.Length.ToString())
        Dim Hex As String = BitConverter.ToString(Data)
        Hex = Hex.Replace("-", "")
        lbLog.Items.Add(Hex)
    End Sub

    Private Sub StringReceived(Sender As Object, Data As String)
        Trace("String received: " + Data)
    End Sub
#End Region

#Region "Get data."
    Private Sub GetByte(Sender As Object)
        CType(Sender, ServerDataProcessor).WriteByte(&HF5)
    End Sub

    Private Sub GetUInt16(Sender As Object)
        CType(Sender, ServerDataProcessor).WriteUInt16(&HF551)
    End Sub

    Private Sub GetUInt32(Sender As Object)
        CType(Sender, ServerDataProcessor).WriteUInt32(&HF5515253)
    End Sub

    Private Sub GetUInt64(Sender As Object)
        CType(Sender, ServerDataProcessor).WriteUInt64(&HF551525354555657)
    End Sub

    Private Sub GetSByte(Sender As Object)
        CType(Sender, ServerDataProcessor).WriteSByte(&HF5)
    End Sub

    Private Sub GetInt16(Sender As Object)
        CType(Sender, ServerDataProcessor).WriteInt16(&HF551)
    End Sub

    Private Sub GetInt32(Sender As Object)
        CType(Sender, ServerDataProcessor).WriteInt32(&HF5515253)
    End Sub

    Private Sub GetInt64(Sender As Object)
        CType(Sender, ServerDataProcessor).WriteInt64(&HF551525354555657)
    End Sub

    Private Sub GetArray(Sender As Object)
        Dim Arr As Byte() = New Byte(511) {}
        For i As UInt16 = 0 To 511
            Arr(i) = wclHelpers.LoByte(i)
        Next
        Dim Res As Int32 = CType(Sender, ServerDataProcessor).WriteArray(Arr)
        If Res <> wclErrors.WCL_E_SUCCESS Then
            Trace("Write failed", Res)
        End If
    End Sub

    Private Sub GetString(Sender As Object)
        CType(Sender, ServerDataProcessor).WriteString("Answer from server")
    End Sub
#End Region
#End Region

#Region "Form events."
    Private Sub fmServerMain_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        FManager = New wclBluetoothManager()
        AddHandler FManager.AfterOpen, AddressOf FManager_AfterOpen
        AddHandler FManager.BeforeClose, AddressOf FManager_BeforeClose

        FServer = New wclRfCommServer()
        ' Disable authentication And encryption.
        FServer.Authentication = False
        FServer.Encryption = False
        ' Use custom service's UUID.
        FServer.Service = ServiceUuid
        AddHandler FServer.OnClosed, AddressOf FServer_OnClosed
        AddHandler FServer.OnConnect, AddressOf FServer_OnConnect
        AddHandler FServer.OnCreateProcessor, AddressOf FServer_OnCreateProcessor
        AddHandler FServer.OnDestroyProcessor, AddressOf FServer_OnDestroyProcessor
        AddHandler FServer.OnDisconnect, AddressOf FServer_OnDisconnect
        AddHandler FServer.OnListen, AddressOf FServer_OnListen

        Dim Res As Int32 = FManager.Open()
        If Res <> wclErrors.WCL_E_SUCCESS Then
            Trace("Bluetooth Manager open failed", Res)
        End If
    End Sub

    Private Sub fmServerMain_FormClosed(sender As Object, e As FormClosedEventArgs) Handles MyBase.FormClosed
        If FServer.State <> wclCommunication.wclServerState.ssClosed Then
            FServer.Close()
        End If
        FServer = Nothing

        If FManager.Active Then
            FManager.Close()
        End If
        FManager = Nothing
    End Sub
#End Region

#Region "Buttons events."
#Region "Clear"
    Private Sub btClear_Click(sender As Object, e As EventArgs) Handles btClear.Click
        lbLog.Items.Clear()
    End Sub
#End Region

#Region "Server buttons"
    Private Sub btListen_Click(sender As Object, e As EventArgs) Handles btListen.Click
        Dim Radio As wclBluetoothRadio = Nothing
        Dim Res As Int32 = FManager.GetClassicRadio(Radio)
        If Res <> wclErrors.WCL_E_SUCCESS Then
            Trace("Get classic radio failed", Res)
        Else
            ' Do Not forget to switch to connectable And discoverable mode!
            Res = Radio.SetConnectable(True)
            If Res <> wclErrors.WCL_E_SUCCESS Then
                Trace("Set connectable failed", Res)
            End If
            Res = Radio.SetDiscoverable(True)
            If Res <> wclErrors.WCL_E_SUCCESS Then
                Trace("Set discoverable failed", Res)
            End If
            Res = FServer.Listen(Radio)
            If Res <> wclErrors.WCL_E_SUCCESS Then
                Trace("Listen failed", Res)
            End If
        End If
    End Sub

    Private Sub btClose_Click(sender As Object, e As EventArgs) Handles btClose.Click
        Dim Res As Int32 = FServer.Close()
        If Res <> wclErrors.WCL_E_SUCCESS Then
            Trace("Close failed", Res)
        End If
    End Sub
#End Region
#End Region
End Class

File: .net/c#/Server/ServerMain.cs
using System;
using System.Windows.Forms;

using wclCommon;
using wclCommunication;
using wclBluetooth;

namespace RfCommServer
{
    public partial class fmServerMain : Form
    {
        private wclBluetoothManager FManager;
        private wclRfCommServer FServer;

        #region Helper methods.
        private void Trace(String str)
        {
            lbLog.Items.Add(str);
        }

        private void Trace(String str, Int32 Error)
        {
            Trace(str + ": 0x" + Error.ToString("X8"));
        }
        #endregion

        #region Bluetooth Manager events.
        private void FManager_BeforeClose(Object sender, EventArgs e)
        {
            Trace("Bluetooth Manager closed");
        }

        private void FManager_AfterOpen(Object sender, EventArgs e)
        {
            Trace("Bluetooth Manager opened.");
        }
        #endregion

        #region Server events.
        private void FServer_OnListen(Object sender, EventArgs e)
        {
            Trace("Server listening");
        }

        private void FServer_OnDisconnect(Object Sender, wclRfCommServerClientConnection Client, Int32 Reason)
        {
            Trace("Client disconnected", Reason);
        }

        private void FServer_OnDestroyProcessor(Object Sender, wclServerClientDataConnection Connection)
        {
            if (Connection.Processor != null)
            {
                Trace("Destroy data processor");
                Connection.Processor.Dispose();
            }
        }

        private void FServer_OnCreateProcessor(Object Sender, wclServerClientDataConnection Connection)
        {
            Trace("Creating data processor");
            ServerDataProcessor Proc = new ServerDataProcessor(Connection);

            Proc.OnByteReceived += Proc_OnByteReceived;
            Proc.OnUInt16Received += Proc_OnUInt16Received;
            Proc.OnUInt32Received += Proc_OnUInt32Received;
            Proc.OnUInt64Received += Proc_OnUInt64Received;
            Proc.OnSByteReceived += Proc_OnSByteReceived;
            Proc.OnInt16Received += Proc_OnInt16Received;
            Proc.OnInt32Received += Proc_OnInt32Received;
            Proc.OnInt64Received += Proc_OnInt64Received;
            Proc.OnArrayReceived += Proc_OnArrayReceived;
            Proc.OnStringReceived += Proc_OnStringReceived;

            Proc.OnGetByte += Proc_OnGetByte;
            Proc.OnGetUInt16 += Proc_OnGetUInt16;
            Proc.OnGetUInt32 += Proc_OnGetUInt32;
            Proc.OnGetUInt64 += Proc_OnGetUInt64;
            Proc.OnGetSByte += Proc_OnGetSByte;
            Proc.OnGetInt16 += Proc_OnGetInt16;
            Proc.OnGetInt32 += Proc_OnGetInt32;
            Proc.OnGetInt64 += Proc_OnGetInt64;
            Proc.OnGetArray += Proc_OnGetArray;
            Proc.OnGetString += Proc_OnGetString;
        }

        private void FServer_OnConnect(Object Sender, wclRfCommServerClientConnection Client, Int32 Error)
        {
            Trace("Client connect", Error);
        }

        private void FServer_OnClosed(Object Sender, Int32 Reason)
        {
            Trace("Server closed");
        }
        #endregion

        #region Data processor events.
        #region Data received.
        private void Proc_OnByteReceived(Object Sender, Byte Data)
        {
            Trace("Byte received: " + Data.ToString("X2") + " (" + Data.ToString() + ")");
        }

        private void Proc_OnUInt16Received(Object Sender, UInt16 Data)
        {
            Trace("UInt16 received: " + Data.ToString("X4") + " (" + Data.ToString() + ")");
        }

        private void Proc_OnUInt32Received(Object Sender, UInt32 Data)
        {
            Trace("UInt32 received: " + Data.ToString("X8") + " (" + Data.ToString() + ")");
        }

        private void Proc_OnUInt64Received(Object Sender, UInt64 Data)
        {
            Trace("UInt64 received: " + Data.ToString("X16") + " (" + Data.ToString() + ")");
        }

        private void Proc_OnSByteReceived(Object Sender, SByte Data)
        {
            Trace("SByte received: " + Data.ToString("X2") + " (" + Data.ToString() + ")");
        }

        private void Proc_OnInt16Received(Object Sender, Int16 Data)
        {
            Trace("Int16 received: " + Data.ToString("X4") + " (" + Data.ToString() + ")");
        }

        private void Proc_OnInt32Received(Object Sender, Int32 Data)
        {
            Trace("Int32 received: " + Data.ToString("X8") + " (" + Data.ToString() + ")");
        }

        private void Proc_OnInt64Received(Object Sender, Int64 Data)
        {
            Trace("Int64 received: " + Data.ToString("X16") + " (" + Data.ToString() + ")");
        }

        private void Proc_OnArrayReceived(Object Sender, Byte[] Data)
        {
            Trace("Array received: " + Data.Length.ToString());
            String Hex = BitConverter.ToString(Data);
            Hex = Hex.Replace("-", "");
            lbLog.Items.Add(Hex);
        }

        private void Proc_OnStringReceived(Object Sender, String Data)
        {
            Trace("String received: " + Data);
        }
        #endregion

        #region Get data.
        private void Proc_OnGetByte(Object Sender)
        {
            ((ServerDataProcessor)Sender).WriteData((Byte)0xF5);
        }

        private void Proc_OnGetUInt16(Object Sender)
        {
            ((ServerDataProcessor)Sender).WriteData((UInt16)0xF551);
        }

        private void Proc_OnGetUInt32(Object Sender)
        {
            ((ServerDataProcessor)Sender).WriteData((UInt32)0xF5515253);
        }

        private void Proc_OnGetUInt64(Object Sender)
        {
            ((ServerDataProcessor)Sender).WriteData((UInt64)0xF551525354555657);
        }

        private void Proc_OnGetSByte(Object Sender)
        {
            ((ServerDataProcessor)Sender).WriteData(unchecked((SByte)0xF5));
        }

        private void Proc_OnGetInt16(Object Sender)
        {
            ((ServerDataProcessor)Sender).WriteData(unchecked((Int16)0xF551));
        }

        private void Proc_OnGetInt32(Object Sender)
        {
            ((ServerDataProcessor)Sender).WriteData(unchecked((Int32)0xF5515253));
        }

        private void Proc_OnGetInt64(Object Sender)
        {
            ((ServerDataProcessor)Sender).WriteData(unchecked((Int64)0xF551525354555657));
        }

        private void Proc_OnGetArray(Object Sender)
        {
            Byte[] Arr = new Byte[512];
            for (UInt16 i = 0; i < 512; i++)
                Arr[i] = wclHelpers.LoByte(i);
            Int32 Res = ((ServerDataProcessor)Sender).WriteData(Arr);
            if (Res != wclErrors.WCL_E_SUCCESS)
                Trace("Write failed", Res);
        }

        private void Proc_OnGetString(Object Sender)
        {
            ((ServerDataProcessor)Sender).WriteData("Answer from server");
        }
        #endregion
        #endregion

        #region Form events.
        private void fmServerMain_Load(Object sender, EventArgs e)
        {
            FManager = new wclBluetoothManager();
            FManager.AfterOpen += FManager_AfterOpen;
            FManager.BeforeClose += FManager_BeforeClose;

            FServer = new wclRfCommServer();
            // Disable authentication and encryption.
            FServer.Authentication = false;
            FServer.Encryption = false;
            // Use custom service's UUID.
            FServer.Service = RfCommClientServer.Commands.ServiceUuid;
            FServer.OnClosed += FServer_OnClosed;
            FServer.OnConnect += FServer_OnConnect;
            FServer.OnCreateProcessor += FServer_OnCreateProcessor;
            FServer.OnDestroyProcessor += FServer_OnDestroyProcessor;
            FServer.OnDisconnect += FServer_OnDisconnect;
            FServer.OnListen += FServer_OnListen;

            Int32 Res = FManager.Open();
            if (Res != wclErrors.WCL_E_SUCCESS)
                Trace("Bluetooth Manager open failed", Res);
        }

        private void fmServerMain_FormClosed(Object sender, FormClosedEventArgs e)
        {
            if (FServer.State != wclCommunication.wclServerState.ssClosed)
                FServer.Close();
            FServer = null;

            if (FManager.Active)
                FManager.Close();
            FManager = null;
        }
        #endregion

        #region Buttons events.
        #region Clear
        private void btClear_Click(Object sender, EventArgs e)
        {
            lbLog.Items.Clear();
        }
        #endregion

        #region Server buttons
        private void btListen_Click(Object sender, EventArgs e)
        {
            wclBluetoothRadio Radio;
            Int32 Res = FManager.GetClassicRadio(out Radio);
            if (Res != wclErrors.WCL_E_SUCCESS)
                Trace("Get classic radio failed", Res);
            else
            {
                // Do not forget to switch to connectable and discoverable mode!
                Res = Radio.SetConnectable(true);
                if (Res != wclErrors.WCL_E_SUCCESS)
                    Trace("Set connectable failed", Res);
                Res = Radio.SetDiscoverable(true);
                if (Res != wclErrors.WCL_E_SUCCESS)
                    Trace("Set discoverable failed", Res);
                Res = FServer.Listen(Radio);
                if (Res != wclErrors.WCL_E_SUCCESS)
                    Trace("Listen failed", Res);
            }
        }

        private void btClose_Click(Object sender, EventArgs e)
        {
            Int32 Res = FServer.Close();
            if (Res != wclErrors.WCL_E_SUCCESS)
                Trace("Close failed", Res);
        }
        #endregion
        #endregion

        public fmServerMain()
        {
            InitializeComponent();
        }
    }
}
